### PR TITLE
Add device activation column to device registry

### PR DESCRIPTION
**Pull request by jochenschneider**

### CODE REVIEW

File: device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala [hunk 1]
```diff
@@ -32,9 +32,10 @@ object DeviceRepository {
     def deviceType = column[DeviceType]("device_type")
     def lastSeen = column[Option[Instant]]("last_seen")
     def createdAt = column[Instant]("created_at")
+    def activatedAt = column[Option[Instant]]("activated_at")
 
-    def * = (namespace, uuid, deviceName, deviceId, deviceType, lastSeen, createdAt).shaped <>
-      ((Device.apply _).tupled, Device.unapply)
+    def * = (namespace, uuid, deviceName, deviceId, deviceType, lastSeen, createdAt, activatedAt)
+      .shaped <> ((Device.apply _).tupled, Device.unapply)
```
Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala]: think we don't need the `.shaped` at all? If it type checks....

File: device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala [hunk 2]
```diff
@@ -105,6 +106,15 @@ object DeviceRepository {
       _ <- devices.insertOrUpdate(newDevice)
     } yield DeviceSeen(device.namespace, uuid, now)
 
+  // inner Future can fail with java.util.NoSuchElementException
```
Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala]: which inner Future?

File: device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala [hunk 3]
```diff
@@ -105,6 +106,15 @@ object DeviceRepository {
...
+  def setActivated(uuid: Uuid)
+                  (implicit ec: ExecutionContext): DBIO[Device] =
+    for {
+      unactivatedDevice <- findByUuid(uuid).filter(_.activatedAt.isEmpty)
+      activatedDevice = unactivatedDevice.copy(activatedAt = Some(Instant.now()))
+      _ <- devices.insertOrUpdate(activatedDevice)
+    } yield activatedDevice
```
Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala]: this should run `.transactionally`

Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala]: You could also run the update, if you have a query to find by uuid AND activatedAt.isEmpty and the `.map` to `activatedAt` and `.update`. Then you could use `handleSingleUpdateError` to check only one row was updated.

If you chose to use the find/update, you could also use `insertOrUpdateWithKey`.

File: device-registry/src/test/scala/org/genivi/sota/device_registry/db/DeviceRepositorySpec.scala [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import org.scalatest.{FunSuite, ShouldMatchers}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+
+class DeviceRepositorySpec extends FunSuite
+                           with DatabaseSpec
+                           with ScalaFutures
+                           with ShouldMatchers {
+
+  test("activatedAt") {
```
Comment (simao) [inline device-registry/src/test/scala/org/genivi/sota/device_registry/db/DeviceRepositorySpec.scala]: Think this should be more specific, ie, "activatedAt is updated after `setActivated`"

**Comment by jochenschneider** [comment]
Reworked with new semantics: set activatedAt is set in updateLastSeen()

File: device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala [hunk 5]
```diff
@@ -97,13 +98,17 @@ object DeviceRepository {
   }
 
   def updateLastSeen(uuid: Uuid)
```
Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala]: IMO this should be written as:

```
  def updateLastSeen(uuid: Uuid)
                    (implicit ec: ExecutionContext): DBIO[Unit] = {

    val now = Instant.now()

    val dbIO = for {
      _ <- devices.filter(_.uuid == uuid).filter(_.activatedAt.isEmpty).map(_.activatedAt).update(Some(now))
      _ <- devices.filter(_.uuid == uuid).map(_.lastSeen).update(Some(now))
    } yield ()

    dbIO.transactionally
  }
```
Also, it should never return a `DeviceSeen`. Device repository is a database encapsulation object, as the  name suggests. DeviceSeen is a BusMessage, a database should not know anything about bus.

Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala]: That also makes sense but also makes sense the repository knows the current time...

Comment (jochenschneider) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala]: The time is needed to construct the BusMessage, so updateLastSeen would either need to return it or take it as a param (I prefer the latter).

Comment (simao) [inline device-registry/src/main/scala/org/genivi/sota/device_registry/db/DeviceRepository.scala]: ah I see! yeah any of those two options look good to me.

**Pull request merged**